### PR TITLE
[MAMAEdu-123048] Saving and transferring parameters of cpa links

### PR DESCRIPTION
**Pull request by Inferato**
**Description:** Saving and transferring parameters of CPA links
It is planned to send traffic via CPA to the "courses showcase" (https://mama-edu.com/courses).
To do this, it is necessary that the parameters in the link from the tracking platform are saved in the browser and transferred to the course landing page.
Required storage time for parameters in the browser: 30 days.

**Youtrack:** https://youtrack.raccoongang.com/issue/MAMAEdu-123048

**AC:**

- The parameters of the CPA link are saved in the browser after clicking on the CPA link. (for the guest and for the authorized user of the lms platform)

- The parameters of the CPA link are saved in the browser after going to other pages of the edX platform or after closing the browser.

- The saved parameters of the CPA link are transferred to the landing page by clicking on the course card or on the "Details" button in the course card.

- CPA-link parameters are ignored (old ones are not overwritten) when re-entering the courses showcase with new CPA-link parameters, provided that the 30-day period of the old parameters has not expired yet.

**Post merge:**
- [ ] Delete working branch (if not needed anymore)

### CODE REVIEW

File: lms/djangoapps/courseware/views/views.py [hunk 1]
```diff
@@ -161,15 +161,30 @@ def courses(request):
...
         "courseware/courses.html",
         {
             'courses': courses_list,
             'course_discovery_meanings': course_discovery_meanings,
-            'programs_list': programs_list
+            'programs_list': programs_list,
         }
     )
 
+    from datetime import timedelta
```
Comment (GlugovGrGlib) [inline lms/djangoapps/courseware/views/views.py:178]: Please move import to the top level imports section.

File: lms/djangoapps/courseware/views/views.py [hunk 2]
```diff
@@ -161,15 +161,30 @@ def courses(request):
...
+    expires = datetime.now() + timedelta(days=30)
+    try:
+        if cpa_cookie:
+            response.set_cookie('cpa_cookie', cpa_cookie, expires=expires)
+    except Exception:
```
Comment (GlugovGrGlib) [inline lms/djangoapps/courseware/views/views.py:183]: Please use a specific Exception type or don't use an exception at all

File: lms/djangoapps/courseware/views/views.py [hunk 3]
```diff
@@ -736,8 +751,18 @@ def course_about(request, course_id):
         if configuration_helpers.get_value('ENABLE_MKTG_SITE', settings.FEATURES.get('ENABLE_MKTG_SITE', False)):
             return redirect(reverse(course_home_url_name(course.id), args=[unicode(course.id)]))
 
+        from ast import literal_eval
+        cpa_cookie = request.COOKIES.get('cpa_cookie', None)
+        try:
+            cpa_cookie = literal_eval(cpa_cookie)
+        except Exception:
```
Comment (GlugovGrGlib) [inline lms/djangoapps/courseware/views/views.py:758]: Please use a specific exception type

File: lms/djangoapps/courseware/views/views.py [hunk 4]
```diff
@@ -736,8 +751,18 @@ def course_about(request, course_id):
         if configuration_helpers.get_value('ENABLE_MKTG_SITE', settings.FEATURES.get('ENABLE_MKTG_SITE', False)):
             return redirect(reverse(course_home_url_name(course.id), args=[unicode(course.id)]))
 
+        from ast import literal_eval
```
Comment (GlugovGrGlib) [inline lms/djangoapps/courseware/views/views.py:754]: please move this import to the top level

Comment (GlugovGrGlib) [reply lms/djangoapps/courseware/views/views.py:754]: @dyudyunov check this ast helpers) it's good to know them) https://docs.python.org/3/library/ast.html#module-ast

File: lms/djangoapps/courseware/views/views.py [hunk 5]
```diff
@@ -736,8 +751,18 @@ def course_about(request, course_id):
...
+        cpa_cookie = request.COOKIES.get('cpa_cookie', None)
+        try:
+            cpa_cookie = literal_eval(cpa_cookie)
+        except Exception:
+            pass
+
+        query_param = ("?" + str(urlencode(cpa_cookie))) if cpa_cookie else None
```
Comment (GlugovGrGlib) [inline lms/djangoapps/courseware/views/views.py:761]: 1. urlencode is not imported in this file. Maybe you want to use urllib.urlencode(cpa_cookie).
2. urllib.urlencode returns a string, so there's no need convert it to str

**Review by GlugovGrGlib — CHANGES_REQUESTED** [review]

File: lms/djangoapps/courseware/views/views.py [hunk 6]
```diff
@@ -161,15 +161,30 @@ def courses(request):
 
     programs_list = get_programs_with_type(include_hidden=False)
 
-    return render_to_response(
+    cpa_cookie = {}
+    if not request.COOKIES.get('cpa_cookie'):
+        for var in request.GET:
+            cpa_cookie[var] = request.GET[var]
```
Comment (GlugovGrGlib) [inline lms/djangoapps/courseware/views/views.py:167]: Interesting, .get() is really slower https://stackoverflow.com/questions/58650208/why-is-key-in-dict-faster-than-dict-getkey-in-python3

**Comment by dyudyunov** [comment]
I have nothing to add to Glib's review. Approved

**Review by GlugovGrGlib — APPROVED** [review]

File: lms/djangoapps/courseware/views/views.py [hunk 7]
```diff
@@ -736,8 +749,18 @@ def course_about(request, course_id):
         if configuration_helpers.get_value('ENABLE_MKTG_SITE', settings.FEATURES.get('ENABLE_MKTG_SITE', False)):
             return redirect(reverse(course_home_url_name(course.id), args=[unicode(course.id)]))
 
+        if 'cpa_cookie' in request.COOKIES:
+            cpa_cookie = request.COOKIES.get('cpa_cookie')
+            try:
+                cpa_cookie = literal_eval(cpa_cookie)
+            except ValueError:
+                pass
```
Comment (dyudyunov) [inline lms/djangoapps/courseware/views/views.py:757]: Is this error handling really useful? If yes, maybe we should at least log the error?

**Review by dyudyunov — APPROVED** [review]

File: lms/djangoapps/courseware/views/views.py (hunk 7 above)
Comment (GlugovGrGlib) [reply lms/djangoapps/courseware/views/views.py:757]: Yes, you're right, it's better to log some useful information to debug easily, but it's not critical there.

**Pull request merged**